File: javascript/largest-series-product/largest-series-product.js
export const largestProduct = (chain, size) => {
  let array = chain.split("").map(el => Number(el));
  let series = [];

  if (chain.match(/[a-zA-Z]\w+/g)) {
    throw new Error("Digits input must only contain digits");
  }
  if (size == 0) {
    return 1;
  }
  if (size < 0) {
    throw new Error("Span must be greater than zero");
  }
  if (chain.length < size) {
    throw new Error("Span must be smaller than string length");
  }

  for (let i = 0; i < array.length; i++) {
    let slice = array.slice(i, i + size);
    if (slice.length == size) {
      series.push(slice);
    }
  }
  const reducer = (accumulator, currentValue) => accumulator * currentValue;

  let products = series.map(el => el.reduce(reducer));
  products.sort((a, b) => {
    return a - b;
  });

  return products[products.length - 1];
};

File: ruby/largest-series-product/largest_series_product.rb
# frozen_string_literal: true

# Class Series
#
class Series
  def initialize(number)
    @number = number
  end

  def largest_product(size)
    return 1 if size.zero?

    valid_number?(size)
    array = []
    @number.chars
           .map(&:to_i)
           .each_cons(size) { |slice| array << slice.reduce(&:*) }
    array.max
  end

  def valid_number?(size)
    if @number.match?(/\D/) ||
       @number.length < size
      raise ArgumentError
    end
  end
end
